### PR TITLE
Fix BadImageFormatException on .NET Core 3.0

### PR DESCRIPTION
**Pull request by lg2de**
This is a PR to provide testing environment for .NET Core for now. We would like to verify e.g. #1047.
Later this PR can be used to add the target for 3.0 in next release.

Regarding double format changes please note https://devblogs.microsoft.com/dotnet/floating-point-parsing-and-formatting-improvements-in-net-core-3-0/

I would recommend to adopt the expectation because this seems to be default behavior in .NET Core 3.

### CODE REVIEW

**Comment by dennisdoomen** [comment]
I think this doesn't work because we still use `msbuild` (through a Cake API).

**Comment by jnyrup** [comment]
Also from `build.cake`
```
settings.ToolVersion = MSBuildToolVersion.VS2017;
```

**Comment by jnyrup** [comment]
And from https://github.com/cake-build/cake/releases there has been some updates from the 0.30.0 that we use to 0.33.0 about NugetRestore and msbuild 16.

**Comment by lg2de** [comment]
With the recent comment the `build.ps1` was running fine.
I guess the build server need to have installed the .NET Core 3 SDK.

Lets retry when  appveyor has finished the VS2019 (16.1) image.

**Comment by jnyrup** [comment]
Could it be that the build server do have 3.0 installed, but since it's preview it's not picked by default?

File: Src/FluentAssertions/Specialized/AssemblyAssertions.cs [hunk 1]
```diff
@@ -19,7 +19,7 @@ public AssemblyAssertions(Assembly assembly)
             Subject = assembly;
         }
 
-#if NET45 || NET47 || NETSTANDARD2_0 || NETCOREAPP2_0 // TODO :: Should be able to remove based on: https://github.com/dotnet/corefx/issues/1784#issuecomment-218803619
```
Comment (lg2de) [inline Src/FluentAssertions/Specialized/AssemblyAssertions.cs]: I tried to remove this, as noted in the TODO.
But I got a lot of errors.
So I recommend to remove now this TODO, because it is not correct (anymore).

File: Src/FluentAssertions/AssertionExtensions.cs [hunk 2]
```diff
@@ -690,7 +690,7 @@ public static FunctionAssertions<T> Should<T>(this Func<T> func)
         }
 
 
-#if NET45 || NET47 || NETCOREAPP2_0
+#if NET45 || NET47 || NETCOREAPP2_0 || NETCOREAPP3_0
```
Comment (lg2de) [inline Src/FluentAssertions/AssertionExtensions.cs]: Should we exclude what is not supported instead include?

**Comment by jnyrup** [comment]
Hmm... I can't really see from the build logs if:
* .net core 3.0 is not installed,
* Restore-NuGet-Packages can't find .net core 3.0, or
* Restore-NuGet-Packages actively chooses .net core 2.2 over .net core 3.0

Raising the logging level from `DotNetCoreVerbosity.Normal` to `DotNetCoreVerbosity.Diagnostic`
could maybe shed some light on those questions.

Maybe related links:
https://help.appveyor.com/discussions/problems/22448-dotnet-core-30-and-netstandard-21
https://github.com/cake-build/cake/issues/2519

File: Src/FluentAssertions/AssertionExtensions.cs (hunk 2 above)
Comment (jnyrup) [reply Src/FluentAssertions/AssertionExtensions.cs]: I agree with this.
Blacklisting instead of whitelisting makes adding new target frameworks easier.

**Comment by dennisdoomen** [comment]
I pushed a commit that switched our Nuget restore and build steps to `dotnet`. The `Visual Studio 2019 Preview` agent has SDK `3.0.100-preview-009812` installed. It now complains about C# issues.

**Comment by jnyrup** [comment]
I don't see any complaints about C# issues?
It fails before finishing restoring nuget packages.

Do we want to add a netcoreapp3.0 target of the Library?
It should suffice to add a netcoreapp3.0 test project.

**Comment by dennisdoomen** [comment]
> I don't see any complaints about C# issues?

I got the entire script running on my local machine, and it even reports the IL errors reported in #1047.

> Do we want to add a netcoreapp3.0 target of the Library?

Not until we need to support assertions on specific types introduced by .NET Core 3.0

**Comment by jnyrup** [comment]
Aha, the #1047 c# issues causing runtime problems.
I thought you were calling out some sort of c# compile time issues.

**Comment by dennisdoomen** [comment]
> I thought you were calling out some sort of c# compile time issues.

As well. But that's probably because that agent is using an old preview of .NET Core 3.0

`C:\projects\fluentassertions\Tests\Shared.Specs\MethodBaseAssertionSpecs.cs(829,77): error CS8107: Feature 'private protected' is not available in C# 7.0. Please use language version 7.2 or greater. `

I'm trying to figure out a way to install the right dotnet SDK as part of our build process. We do this as well in our projects.

File: Src/FluentAssertions/Specialized/AssemblyAssertions.cs (hunk 1 above)
Comment (jnyrup) [reply Src/FluentAssertions/Specialized/AssemblyAssertions.cs]: @dennisdoomen Do you have an opinion to this?

**Comment by jnyrup** [comment]
@lg2de As #1060 has been merged it should now be possible to build .net core 3.0 targets.
As we're not utilizing any .net core 3.0 features (yet!), please remove that from the core library target frameworks.
When I tried locally I could reproduce #1047 with a netcoreapp2.0 core library and a netcoreapp3.0 test project.

File: Src/FluentAssertions/Specialized/AssemblyAssertions.cs (hunk 1 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/Specialized/AssemblyAssertions.cs]: If we (still) can't remove, don't

**Comment by lg2de** [comment]
I have reworked as requested and added the fix mentioned by @jnyrup.
Please review my changes and comment the TODOs added.
The build failed now "only" due to timing issues.

File: Tests/Benchmarks/Benchmarks.csproj [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFramework>net471</TargetFramework>
+    <TargetFramework>net47</TargetFramework>
```
Comment (dennisdoomen) [inline Tests/Benchmarks/Benchmarks.csproj]: Why is this needed?

File: Src/FluentAssertions/FluentAssertions.csproj [hunk 4]
```diff
@@ -14,7 +14,7 @@
     <Authors>Dennis Doomen;Jonas Nyrup</Authors>
     <PackageDescription>
       A very extensive set of extension methods that allow you to more naturally specify the expected outcome of a TDD or
-      BDD-style unit tests. Targets .NET Framework 4.5 and 4.7, as well as .NET Core 2.0, .NET Standard 1.3, 1.6 and 2.0.
+      BDD-style unit tests. Targets .NET Framework 4.5 and 4.7, as well as .NET Core 2.0, .NET Core 3.0, .NET Standard 1.3, 1.6 and 2.0.
```
Comment (jnyrup) [inline Src/FluentAssertions/FluentAssertions.csproj]: We're no longer targeting net core 3.0

File: Src/FluentAssertions/FluentAssertions.csproj [hunk 5]
```diff
@@ -56,6 +56,9 @@
   <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp2.0'">
     <PackageReference Include="System.Configuration.ConfigurationManager" Version="4.4.0" />
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp3.0'">
```
Comment (jnyrup) [inline Src/FluentAssertions/FluentAssertions.csproj]: We're no longer targeting net core 3.0

File: build.cake [hunk 6]
```diff
@@ -50,7 +50,6 @@ Task("Restore-NuGet-Packages")
 	DotNetCoreRestore(new DotNetCoreRestoreSettings
 	{
 		NoCache = true,
-        ConfigFile = "./nuget.config",
```
Comment (jnyrup) [inline build.cake]: Please keep this

File: Tests/Shared.Specs/Utilities.cs [hunk 7]
```diff
@@ -2,7 +2,10 @@
 
 namespace FluentAssertions.Specs
 {
-    public static class Utilities
+    /// <summary>
+    /// Implements extensions to <see cref="TimeSpan"/> available in .NET Core 2, but not in .NET Framework.
+    /// </summary>
+    public static class Utilities // TODO Rename to TimeSpanExtensions?
```
Comment (jnyrup) [inline Tests/Shared.Specs/Utilities.cs]: Good idea, that's a way better name than my `Utilities`

File: Tests/Shared.Specs/EventAssertionSpecs.cs [hunk 8]
```diff
@@ -744,7 +744,7 @@ public void When_monitoring_an_object_through_an_interface_it_should_monitor_onl
             }
         }
 
-#if NET45 || NET47
+#if NET45 || NET47 // TODO Why this is not available in e.g. core2?
```
Comment (jnyrup) [inline Tests/Shared.Specs/EventAssertionSpecs.cs]: If you add `NETCOREAPP2_0` it does not compile, as `DefineDynamicAssembly` and `DefineDynamicModule` are not defined for that target.

File: Src/FluentAssertions/FluentAssertions.csproj (hunk 5 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/FluentAssertions.csproj]: But we do support it officially. So I would keep it in.

Comment (jnyrup) [reply Src/FluentAssertions/FluentAssertions.csproj]: Isn't this superfluous if we aren't targeting net core 3.0?

Comment (dennisdoomen) [reply Src/FluentAssertions/FluentAssertions.csproj]: Correct.

File: Src/FluentAssertions/FluentAssertions.csproj (hunk 4 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/FluentAssertions.csproj]: But we do support it officially. So I would keep it in.

File: Src/FluentAssertions/FluentAssertions.csproj (hunk 5 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/FluentAssertions.csproj]: My response was meant for your comment on the nuspec file.

File: Src/FluentAssertions/FluentAssertions.csproj (hunk 4 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/FluentAssertions.csproj]: Maybe the PR can also update the `docs` folder.

File: Tests/Benchmarks/Benchmarks.csproj (hunk 3 above)
Comment (lg2de) [reply Tests/Benchmarks/Benchmarks.csproj]: This is not needed, but I would recommend this for consistency of the solution.
The target net47 is also build for 4.7.0.
Currently you need to install SDK for 4.7.0 **and** 4.7.1 to build the solution.

File: Src/FluentAssertions/FluentAssertions.csproj (hunk 4 above)
Comment (lg2de) [reply Src/FluentAssertions/FluentAssertions.csproj]: Done.
Please check.

File: build.cake (hunk 6 above)
Comment (lg2de) [reply build.cake]: This was a mistake due to merge from master.
Reverted.
